src/coreComponents/mesh/generators/PAMELAMeshGenerator.cpp

```
@@ -61,319 +56,435 @@ PAMELAMeshGenerator::PAMELAMeshGenerator( string const & name, Group * const par
+    string const & srcRegionName = m_cellBlockRegions.at( subRegion.getName() );
+    PAMELA::Part< PAMELA::Polyhedron * > * const regionPtr = polyhedronPartMap.at( srcRegionName );
+    GEOSX_ASSERT( regionPtr != nullptr );
+
+    // Use element type to get the PAMELA cell block (sub-part) corresponding to subregion
+    PAMELA::ELEMENTS::TYPE const elemType = toPamelaElementType( subRegion.getElementType() );
+    PAMELA::SubPart< PAMELA::Polyhedron * > * const cellBlockPtr = regionPtr->SubParts.at( static_cast< int >( elemType ) );
+    GEOSX_ASSERT( cellBlockPtr != nullptr );
 
-    string surfaceName = DecodePAMELALabels::retrieveSurfaceOrRegionName( surfacePtr->Label );
-    SortedArray< localIndex > & curNodeSet  = nodeSets.registerWrapper< SortedArray< localIndex > >( string( surfaceName ) ).reference();
-    for( auto const & subPart : surfacePtr->SubParts )
+    // Make a list of material field wrapper names on current subregion
+    std::unordered_set< string > const materialWrapperNames = getMaterialWrapperNames( subRegion );
+
+    for( localIndex fieldIndex = 0; fieldIndex < m_fieldsToImport.size(); fieldIndex++ )
     {
-      auto const cellBlockPAMELA = subPart.second;
-      PAMELA::ELEMENTS::TYPE const cellBlockType = cellBlockPAMELA->ElementType;
-      string const cellBlockName = ElementToLabel.at( cellBlockType );
-      if( cellBlockName == "TRIANGLE"  || cellBlockName == "QUAD" )
+      string const & sourceName = m_fieldsToImport[fieldIndex];
+      string const & wrapperName = m_fieldNamesInGEOSX[fieldIndex];
+
+      // Find source
+      PAMELA::VariableDouble * const meshProperty = regionPtr->FindVariableByName( sourceName );
+      GEOSX_ASSERT( meshProperty != nullptr );
+
+      // Find destination
+      WrapperBase & wrapper = subRegion.getWrapperBase( wrapperName );
+
+      // Decide if field is constitutive or not
+      if( materialWrapperNames.count( wrapperName ) > 0 )
```

Do we maybe want to throw an error if the property does not already exist in GEOSX?

---

`getWrapperBase()` will throw an error here, do you mean throwing a more user-friendly error? That sounds like a good idea.

---

Yeah, I meant like a more user-friendly error than what comes out of `getWrapperBase()`.